### ISSUE TITLE
Joined pound in function name parse failed

### ISSUE THREAD

**Issue by A4-Tacks** [koalaman/shellcheck#3102]
#### For bugs
- Rule Id: SC1073
- My shellcheck version (`shellcheck --version` or "online"):
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/bash

function x#y {
    echo foo
}

x#y
```

#### Here's what shellcheck currently says:
```

In - line 3:
function x#y {
^-- SC1073 (error): Couldn't parse this function. Fix to allow more checks.


In - line 4:
    echo foo
    ^-- SC1064 (error): Expected a { to open the function definition.
    ^-- SC1072 (error):  Fix any mentioned problems and try again.

For more information:
  https://www.shellcheck.net/wiki/SC1064 -- Expected a { to open the function...
  https://www.shellcheck.net/wiki/SC1072 --  Fix any mentioned problems and t...
  https://www.shellcheck.net/wiki/SC1073 -- Couldn't parse this function. Fix...
```

#### Here's what I wanted or expected to see:
no outputs
